src/zkp/pienc.rs

```diff
@@ -279,27 +275,204 @@ mod tests {
+            setup_params,
+            encryption_key: encryption_key.clone(),
+            ciphertext,
+        };
+
+        // Correctly formed proof verifies correctly
+        let proof = PiEncProof::prove(
+            rng,
+            &input,
+            &PiEncSecret {
+                plaintext: plaintext.clone(),
+                nonce: nonce.clone(),
+            },
+        )?;
+        assert!(proof.verify(&input).is_ok());
+
+        // Forming with the wrong plaintext fails
+        let wrong_plaintext = random_plusminus_by_size(rng, ELL);
+        assert_ne!(wrong_plaintext, plaintext);
+        let proof = PiEncProof::prove(
+            rng,
+            &input,
+            &PiEncSecret {
+                plaintext: wrong_plaintext,
+                nonce: nonce.clone(),
+            },
+        )?;
+        assert!(proof.verify(&input).is_err());
+
+        // Forming with the wrong nonce fails
+        let (_, wrong_nonce) = encryption_key.encrypt(rng, &plaintext)?;
+        assert_ne!(wrong_nonce, nonce);
+        let proof = PiEncProof::prove(
+            rng,
+            &input,
+            &PiEncSecret {
+                plaintext,
+                nonce: wrong_nonce,
+            },
+        )?;
+        assert!(proof.verify(&input).is_err());
+
+        Ok(())
+    }
+
+    #[test]
+    fn verification_requires_correct_common_input() -> Result<()> {
+        // Replace each field of `CommonInput` with something else to verify
+        let mut rng = crate::utils::get_test_rng();
 
-        // A plaintext that's in range for encryption but larger than 2^ELL should fail
-        assert!(random_paillier_encryption_in_range_proof(&mut rng, too_large).is_err());
+        // Form inputs
+        let plaintext = random_plusminus_by_size(&mut rng, ELL);
+        let (proof, mut input) = build_random_proof(&mut rng, plaintext.clone())?;
+
+        // Verification works on the original input
+        assert!(proof.verify(&input).is_ok());
+
+        // Verification fails with the wrong setup params
+        let (bad_decryption_key, _, _) = DecryptionKey::new(&mut rng)?;
+        let bad_setup_params = VerifiedRingPedersen::extract(&bad_decryption_key, &mut rng)?;
+        let setup_params = input.setup_params;
+        input.setup_params = bad_setup_params;
+        assert!(proof.verify(&input).is_err());
+        input.setup_params = setup_params;
+
+        // Verification fails with the wrong encryption key
+        let bad_encryption_key = DecryptionKey::new(&mut rng)?.0.encryption_key();
+        let encryption_key = input.encryption_key;
+        input.encryption_key = bad_encryption_key;
+        assert!(proof.verify(&input).is_err());
+        input.encryption_key = encryption_key;
+
+        // Verification fails with the wrong ciphertext
+        let (bad_ciphertext, _) = input.encryption_key.encrypt(&mut rng, &plaintext)?;
+        input.ciphertext = bad_ciphertext;
+        assert!(proof.verify(&input).is_err());
```

Comment:
It might be good to save the ciphertext and reset things to avoid any future headaches if someone adds a test to the end and doesn't realize `input` has not been reset.

Reply:
Rewrote this a bit to match the formatting you used in #151, so this should be an issue any more!